pip/pip-323.md

```diff
@@ -0,0 +1,167 @@
+topic stats endpoints (`{tenant}/{namespace}/{topic}/stats` and `{tenant}/{namespace}/{topic}/partitioned-stats`),
+allowing the user a complete workflow: alert using metrics when topic backlog is about to be exceeded, then
+query topic stats for that topic to retrieve the subscription name which contains the oldest message.
+For completeness, we will also add the backlog quota limits, both age and size, and the age of oldest 
+unacknowledged message.
+
+We will add a metric allowing the user to know how many times the usage exceeded the quota, both for time or size -
+`pulsar_storage_backlog_quota_exceeded_evictions_total`, where the `quota_type` label will be either `time` or 
+`size`. Monitoring that counter over time will allow the user to know when a topic backlog exceeded its quota,
+and if backlog eviction was chosen as action, then it happened, and how many times. 
+
+Some users may want the backlog quota check to happen more frequently, and as a consequence, the backlog age 
+metric more frequently updated. They can modify `backlogQuotaCheckIntervalInSeconds` configuration key, but without
+knowing how long this check takes, it will be hard for them. Hence, we will add the metric
+`pulsar_storage_backlog_quota_check_duration_seconds` which will be of histogram type.
+
+# Detailed Design
+
+## Public-facing Changes
+
+### Public API
+Adding the following to the response of topic stats, of both `{tenant}/{namespace}/{topic}/stats` 
+and `{tenant}/{namespace}/{topic}/partitioned-stats`:
+
+* `backlogQuotaLimitSize` - the size in bytes of the topic backlog quota 
+* `backlogQuotaLimitTime` - the topic backlog age quota, in seconds. 
+* `oldestBacklogMessageAgeSeconds` - the age of the oldest unacknowledged (i.e. backlog) message, measured by 
+   the time elapsed from its published time, in seconds. This value is recorded every backlog quota check 
+   interval, hence it represents the value seen in the last check.
+* `oldestBacklogMessageSubscriptionName` - the name of the subscription containing the oldest unacknowledged message.
+  This value is recorded every backlog quota check interval, hence it represents the value seen in the last check.
+
+
+### Metrics
+
+| Name                                                    | Description                                                                        | Attributes                        | Units   |
+|---------------------------------------------------------|------------------------------------------------------------------------------------|-----------------------------------|---------|
+| `pulsar_storage_backlog_age_seconds`                    | Gauge. The age of the oldest unacknowledged message (backlog)                      | cluster, topic                    | seconds |
+| `pulsar_storage_backlog_quota_exceeded_evictions_total` | Counter. The number of times a backlog was evicted since it has exceeded its quota | cluster, topic, quota_type = time |         | 
```

Comment:
Do we need to consider adding broker-level and namespace-level metrics?

- pulsar_broker_storage_backlog_evictions_total (cluster)
- pulsar_ns_storage_backlog_evictions_total (cluster, namespace)
- pulsar_topic_storage_backlog_evictions_total (cluster, namespace, topic)

Reply:
Actually, I don't remember why we had `storage` for `pulsar_storage_backlog_size` before.
Now, I don't think it will help something

"pulsar_backlog_size" also looks good.

Similarly, maybe we can consider to use

- pulsar_broker_backlog_evictions_total (cluster)
- pulsar_ns_backlog_evictions_total (cluster, namespace)
- pulsar_topic_backlog_evictions_total (cluster, namespace, topic)

Reply:
Regarding `pulsar_broker_storage_backlog_evictions_total`. The current design is to either emit it as topic of namespace level, with the same name, so I don't want to deviate from that.
Regarding broker level: Good call - I completely forgot about it. Added `pulsar_broker_storage_backlog_quota_exceeded_evictions_total`